bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,193 @@
+static QUEUES: Lazy<RwLock<HashMap<Backend, Vec<String>>>> =
+    Lazy::new(|| RwLock::new(HashMap::new()));
+
+#[async_trait]
+impl ProxyHttp for LB {
+    type CTX = ();
+
+    fn new_ctx(&self) {}
+
+    async fn upstream_peer(
+        &self,
+        session: &mut Session,
+        _ctx: &mut Self::CTX,
+    ) -> Result<Box<HttpPeer>> {
+        let upstream = self.0.select(b"", 256).unwrap();
+
+        // read request ID from headers
+        let request_id = session
+            .get_header("X-Request-ID")
+            .expect("Request ID not found")
+            .to_str()
+            .expect("Invalid header value");
+
+        {
+            let mut ctx_guard = QUEUES.write().await;
+            let backend_queue = ctx_guard.entry(upstream.clone()).or_insert_with(Vec::new);
+
+            // Limit queue length to MAX_QUEUE_ITEMS requests
+            if backend_queue.len() >= MAX_QUEUE_ITEMS {
+                panic!("Too many requests in the queue");
+                // return Err(Box::new("Too many requests in the queue".into()));
+            }
+
+            backend_queue.push(request_id.to_string());
+        }
+
+        // Wait for the request to be at the front of the queue
+        loop {
+            // We use a new scope for each iteration to release the lock
+            {
+                let ctx_guard = QUEUES.read().await;
+                if let Some(backend_queue) = ctx_guard.get(&upstream) {
+                    if backend_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    // TODO: replace this panic with an error
+                    panic!("Upstream not found");
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // Set SNI
+        let mut http_peer = HttpPeer::new(upstream, false, "".to_string());
+        let peer_opts = http_peer.get_mut_peer_options().unwrap();
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = TIMEOUT_SECS;
+        peer_opts.read_timeout = TIMEOUT_SECS;
+        peer_opts.write_timeout = TIMEOUT_SECS;
+        peer_opts.idle_timeout = TIMEOUT_SECS;
+
+        peer_opts.alpn = ALPN::H2;
+
+        let peer = Box::new(http_peer);
+        Ok(peer)
+    }
+
+    async fn upstream_request_filter(
+        &self,
+        _session: &mut Session,
+        upstream_request: &mut RequestHeader,
+        _ctx: &mut Self::CTX,
+    ) -> Result<()>
+    where
+        Self::CTX: Send + Sync,
+    {
+        // Check if it's a gRPC request
+        if let Some(content_type) = upstream_request.headers.get("content-type") {
+            if content_type == "application/grpc" {
+                // Ensure the correct host and gRPC headers are forwarded
+                // upstream_request.insert_header("Host", "0.0.0.0").unwrap();
```

Comment:
I think you forgot to erase this debug comment

Reply:
Thanks!